### ISSUE TITLE
How does sourmash gather time scale with reads (and can this be reduced with multithreading)?

### ISSUE THREAD

**Issue by rmcolq** [sourmash-bio/sourmash#2827]
I'd really like to use sourmash for metagenomic classification as in portik et al. I have been trying it out on small datasets and I've noticed that the gather step in particular seems to take a long time (e.g. it took 6.5 mins to process 40k long reads, whilst kraken2 run with a server took 4.5s on the same data). I will be running classification in an environment where there is lots of access to RAM so holding an ~80GB database in memory is not a problem, but as we anticipate having to process lots of samples we need the time to be as short as possible.

Do you plan to introduce multithreading to sourmash gather?
Does this time scale with the number of reads, or is it mostly the time to load the database?
Is there something I could change e.g. database or threshold_bps which would speed things up?

**Comment by ctb** [comment]
hi @rmcolq, thanks for asking!

I have a couple of thoughts for you!

First, we have a multithreaded gather implemented over in [pyo3_branchwater](https://github.com/sourmash-bio/pyo3_branchwater). We tend to use it in two stages - run fastgather, and then use the fastgather results [as a picklist](https://github.com/sourmash-bio/pyo3_branchwater/blob/main/doc/README.md#using-fastgather-to-create-a-picklist-for-sourmash-gather) to run a regular gather to do more complete reporting. Happy to describe this more if you try it out and need some help; the hardest part is getting it installed, I'm afraid, but once that's done it should be smooth sailing.

@bluegenes recently gave a lab meeting on branchwater; see [slides here](https://drive.google.com/file/d/1PbRPzKUCAJEyTlSKYahb8kKgWp8LG1uD/view).

In particular, the fastmultigather command might be what you need for many samples. It loads the database once and then processes all of the samples in parallel.

---

Second, we rarely hear "sure, hold 80 GB in memory!" so I have to admit I'm not 100% sure what to recommend 😆 

But what you're seeing is what I'd expect: gather does a pass across the full database (the prefetch stage) before reporting matches. So if you are using (e.g.) the entire GTDB, sourmash does a pass across the entire database, loading it from disk as it goes. Hence the ~5 minutes.

---

Third, there are a couple of indexed databases that might be faster. My first suggestion is to try out the sqlite index (`sourmash sig cat database -o database.sqldb`); my second suggestion is to try out the SBT index (generated with `sourmash index`. Neither will get you to 5 seconds, though. The mastiff index (in branchwater) might, but if you are using branchwater, I think fastmultigather is your best bet.

---

A few other thoughts for us sourmash devs to think about -

* if you're using a server, kraken is pre-loading the database, presumably. So that's one reason why it's so darn fast. We don't have that implemented in sourmash directly because we have focused on keeping memory usage low (that's more often a limiting factor).
* sourmash can deal with much larger databases than kraken, which is one reason why we've focused on keeping RAM low (at expense of time).

**Comment by rmcolq** [comment]
Hi @ctb 

Thank you for this in depth answer! It sounds like fastmultigather gets a long way towards what we are looking for, so I'll definitely check it out. I have been using the largest databases on the sourmash downloads page although that 6 minutes was just with the viral, fungi and protozoa files for the test data (I was planning to use bacteria too in the main). Makes sense that the focus would be on getting the RAM use down. 

The fact that most of that time is in database loading is pretty reassuring because it means that something like the multithreading one (or a server implementation of it if it were to occur down the line) would be a good solution.

**Comment by ctb** [comment]
I've written a short `fastmultigather` quickstart here: https://github.com/sourmash-bio/sourmash/issues/3095. This should dramatically decrease loading time as well as memory usage.

benchmarks:
* [indexing 300k GTDB genomes @ 5 hours, 14 GB of RAM](https://github.com/sourmash-bio/sourmash_plugin_branchwater/issues/214#issuecomment-1948393314)
* [full metagenome gather @ 2m, 600 MB RAM](https://github.com/sourmash-bio/sourmash_plugin_branchwater/issues/214#issuecomment-1948414114)

**Comment by ctb** [comment]
note: as of sourmash_plugin_branchwater v0.9.5 [link](https://github.com/sourmash-bio/sourmash_plugin_branchwater/releases/tag/v0.9.5), the results from `fastgather` and `fastmultigather` are now identical to those from `sourmash gather`. So we can just recommend using `fastmultigather` directly.

I haven't done the appropriate benchmarking yet, but I'll get to it eventually :).

**Comment by rmcolq** [comment]
Amazing, thanks for keeping this updated with the developments!

**Comment by ctb** [comment]
Here are the benchmarks - in brief, `fastgather` is quite fast, while the rocksdb implementation is very low memory:

https://github.com/sourmash-bio/sourmash/issues/3232

**Comment by ezherman** [comment]
> Here are the benchmarks - in brief, `fastgather` is quite fast, while the rocksdb implementation is very low memory:
> 
> [#3232](https://github.com/sourmash-bio/sourmash/issues/3232)

Hi! I was wondering whether `fastgather` is still expected to be much faster than `gather` when using a `RocksDB` database with the latest version of sourmash. I'm happy to open a separate issue if that's more appropriate.

Context: we're looking to use `picklist`, however this is not currently supported in `fastgather`.

**Comment by ctb** [comment]
Great question!  The latest benchmarks are actually in this PR, https://github.com/sourmash-bio/sourmash/pull/3545, and this should reflect current performance (August 2025, sourmash v4.9.4, branchwater plugin v0.9.13). I put a screenshot of the numbers from that PR at the bottom of this comment!

tl;dr RocksDB with regular `sourmash gather` would be my recommendation (and is what I usually use).

Details:

* `sourmash gather` against a RocksDB is not _quite_ as fast as `fastmultigather` against a RocksDB, but it's still ridiculously fast and low memory;
* `fastgather` doesn't support RocksDB - only `fastmultigather` does;

The one caveat I have is that we have not done any serious benchmarking with picklists and RocksDB... so there may be some hidden 🐲 there. So, umm, please loop us in if you notice serious performance problems 😆 . Ref https://github.com/sourmash-bio/sourmash/issues/3592

HTH!

<img width="1090" height="828" alt="Image" src="https://github.com/user-attachments/assets/4d30b865-c6bf-4136-b101-77278656ef57" />

**Comment by ezherman** [comment]
Amazing, thank you @ctb!

> The one caveat I have is that we have not done any serious benchmarking with picklists and RocksDB... so there may be some hidden 🐲 there. So, umm, please loop us in if you notice serious performance problems 😆 . Ref https://github.com/sourmash-bio/sourmash/issues/3592

Alright, you'll hear from us if we run into any issues 😁
